internal/cmd/update.go

```diff
@@ -0,0 +1,466 @@
+package cmd
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"cdr.dev/coder-cli/internal/config"
+	"cdr.dev/coder-cli/internal/version"
+	"cdr.dev/coder-cli/pkg/clog"
+	"golang.org/x/xerrors"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/manifoldco/promptui"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+const (
+	goosWindows       = "windows"
+	goosLinux         = "linux"
+	apiPrivateVersion = "/api/private/version"
+)
+
+// updater updates coder-cli.
+type updater struct {
+	confirmF       func(string) (string, error)
+	execF          func(context.Context, string, ...string) ([]byte, error)
+	executablePath string
+	fs             afero.Fs
+	httpClient     getter
+	osF            func() string
+	versionF       func() string
+}
+
+func updateCmd() *cobra.Command {
+	var (
+		force      bool
+		coderURL   string
+		versionArg string
+	)
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "Update coder binary",
+		Long:  "Update coder to the version matching a given coder instance.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			httpClient := &http.Client{
+				Timeout: 10 * time.Second,
+			}
+
+			currExe, err := os.Executable()
+			if err != nil {
+				return clog.Fatal("init: get current executable", clog.Causef(err.Error()))
+			}
+
+			updater := &updater{
+				confirmF:       defaultConfirm,
+				execF:          defaultExec,
+				executablePath: currExe,
+				httpClient:     httpClient,
+				fs:             afero.NewOsFs(),
+				osF:            func() string { return runtime.GOOS },
+				versionF:       func() string { return version.Version },
+			}
+			return updater.Run(ctx, force, coderURL, versionArg)
+		},
+	}
+
+	cmd.Flags().BoolVar(&force, "force", false, "do not prompt for confirmation")
+	cmd.Flags().StringVar(&coderURL, "coder", "", "query this coder instance for the matching version")
+	cmd.Flags().StringVar(&versionArg, "version", "", "explicitly specify which version to fetch and install")
+
+	return cmd
+}
+
+type getter interface {
+	Get(url string) (*http.Response, error)
+}
+
+func (u *updater) Run(ctx context.Context, force bool, coderURLArg string, versionArg string) error {
```

Comment:
This function could benefit from being split up a little bit more, it's fairly long as it stands right now